components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -93,6 +94,37 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes here.
+    protected readonly contentCache = new Map<string, string>();
+    protected readonly updatedSinceLastWrite: string[] = [];
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.push(resource.uri);
+    }
+
+    protected async cachedRead(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+        return this.getServer().getUserStorageResource(req);
```

Comment:
the result should be written to the cache

Reply:
We decided against it as we assumed that no meaningful changes happen outside the workspace. It's a race anyways.

Reply:
this way it stands now, all reads are made to the server until the first write happens. Is that intentional?

Reply:
Yes, I thought that would be the conclusion we had in our meeting...? We want to guard against too many _writes_.

Reply:
Added it due to popular demand :upside_down_face:

Reply:
The meetings conclusion was imho to not re-sync the cache from the server and try and merge with previous writes.
I reckon that it's fine to initialise the cache with a version read from the server if we haven't written that one yet.